pandas/core/frame.py

```diff
@@ -3700,17 +3709,20 @@ def combine(self, other, func, fill_value=None, overwrite=True):
                 otherSeries[other_mask] = fill_value
 
             # if we have different dtypes, possibily promote
-            new_dtype = this_dtype
-            if this_dtype != other_dtype:
-                new_dtype = _lcd_dtypes(this_dtype, other_dtype)
-                series = series.astype(new_dtype)
+            if notnull(series).all():
+                new_dtype = this_dtype
```

Comment:
in another life ``.combine`` functionaility could be pushed to block-manager (there *is* a combine there already, but it has a different purpose)